### PR TITLE
refactor: split general request into general unknown request and general action and label request (PL-1124)

### PR DESCRIPTION
**Pull request by zhihil**
**Fixes or implements PL-1124**

### Brief description. What is this change?

As a result of this [investigation](https://www.notion.so/voiceflow/Malformed-General-Runtime-Requests-d990f137c2dd47398ec58cc8f7ec479b?pvs=4), we've identified malformed requests sent by users. One class of these errors includes general requests whose payloads can be anything.

In this PR, we loosen the type of the general request in `libs` to reflect how we've loosened the types in `@voiceflow/dtos`.

### Implementation details. How do you make this change?

Split the GeneralRequest type into two sub-types GeneralUnknownRequest and GeneralActionAndLabelRequest.

GeneralActionAndLabelRequest is more or less what GeneralRequest is currently implemented as.

GeneralUnknownRequest is a new type where payload is unknown, which is intended to accommodate the myriad of data types users could insert for payload rather than a ActionAndLabelRequestPayload.

### Setup information

N/A

### Deployment Notes

N/A

### Related PRs

1. https://github.com/voiceflow/creator-app/pull/8215
2. https://github.com/voiceflow/libs/pull/527
3. https://github.com/voiceflow/general-runtime/pull/837

### Checklist

- ~~[ ] Breaking changes have been communicated, including:~~
    - New required environment variables
    - Renaming of interfaces (API routes, request/response interface, etc)
- ~~[ ] New environment variables have [been deployed](https://www.notion.so/voiceflow/Add-Environment-Variables-be1b0136479f45f1adece7995a7adbfb)~~
- ~~[ ] Appropriate tests have been written~~
    - Bug fixes are accompanied by an updated or new test
    - New features are accompanied by a new test

### CODE REVIEW

File: packages/base-types/sonar-project.properties [hunk 1]
```diff
@@ -1,3 +1,8 @@
-sonar.projectName=libs-base-types
+sonar.projectName=libs-metrics
```
Comment (zhihil) [inline packages/base-types/sonar-project.properties:1]: Enabled unit tests to pass SonarCloud.

File: packages/base-types/src/request/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,312 @@
+import { describe, expect, it } from 'vitest';
```
Comment (zhihil) [inline packages/base-types/src/request/index.test.ts:1]: Added unit testing to ensure SonarCloud passes this PR.

File: packages/base-types/src/request/index.ts [hunk 3]
```diff
@@ -80,9 +80,20 @@ export interface IntentRequest extends BaseRequest {
...
   type: string; // the general request type is dynamic
-  payload?: ActionAndLabelRequestPayload;
+  payload?: unknown;
+}
+
+export interface GeneralActionAndLabelRequest extends GeneralUnknownRequest {
+  payload: ActionAndLabelRequestPayload;
+}
+
+export type GeneralRequest = GeneralUnknownRequest | GeneralActionAndLabelRequest;
```
Comment (zhihil) [inline packages/base-types/src/request/index.ts:92]: Split `GeneralRequest` into two subtypes, `GeneralUnknownRequest` and `GeneralActionAndLabelRequest` where the former type reflects the actual request data we are now accounting for.

File: packages/base-types/src/request/index.ts [hunk 4]
```diff
@@ -114,6 +125,20 @@ export interface NodeButton {
...
+  request.payload !== null &&
+  'actions' in request.payload &&
+  Array.isArray(request.payload.actions);
+
+export const hasLabelPayload = (request: BaseRequest): request is { type: string; payload: LabelRequestPayload } =>
+  'payload' in request &&
+  typeof request.payload === 'object' &&
+  request.payload !== null &&
+  'label' in request.payload &&
+  typeof request.payload.label === 'string';
```
Comment (zhihil) [inline packages/base-types/src/request/index.ts:140]: These type-guards are useful to resolve type issues in `general-runtime`. 

Some code needs to access `request.payload.actions` or `request.payload.label` but we don't ultimately care what type of request it is, so long as `.actions` and `.label` are defined properly. 

Therefore, these type-guards are useful to detecting partial request types.

File: packages/base-types/tsconfig.build.json [hunk 5]
```diff
@@ -1,4 +1,5 @@
 {
   "extends": "./tsconfig.json",
-  "include": ["src"]
+  "include": ["src"],
+  "exclude": ["**/*.test.ts"]
```
Comment (zhihil) [inline packages/base-types/tsconfig.build.json:4]: Ignore test files for the build.

File: packages/base-types/vitest.config.mts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line no-restricted-exports
+export { default } from '@voiceflow/vitest-config';
```
Comment (zhihil) [inline packages/base-types/vitest.config.mts:2]: Setup `vitest` for unit testing this repository.

**Review by zhihil — COMMENTED** [review]
For the reader's understanding:

File: packages/base-types/sonar-project.properties (hunk 1 above)
Comment (theProf) [reply packages/base-types/sonar-project.properties:1]: what is this change actually doing?

**Review by trs — APPROVED** [review]

**Comment by DecathectZero** [comment]
we're going to use `@voiceflow/dtos-interact` as our main source of truth, and avoid `libs` changes in the future.

**Pull request closed without merging**